c10/util/intrusive_ptr.h

```diff
@@ -686,23 +695,25 @@ using weak_intrusive_ptr_target = intrusive_ptr_target;
 namespace raw {
 
 namespace intrusive_ptr {
-
-  // WARNING: Unlike the reclaim() API, it is NOT valid to pass
-  // NullType::singleton to this function
   inline void incref(intrusive_ptr_target* self) {
-    auto ptr = c10::intrusive_ptr<intrusive_ptr_target>::reclaim(self);
-    auto ptr_copy = ptr;
-    ptr_copy.release();
-    ptr.release();
+    if (self) {
+      ++self->refcount_;
+    }
   }
 
-  // WARNING: Unlike the reclaim() API, it is NOT valid to pass
-  // NullType::singleton to this function
   inline void decref(intrusive_ptr_target* self) {
-    // Let it die
-    c10::intrusive_ptr<intrusive_ptr_target>::reclaim(self);
-    // NB: Caller still has 'self' pointer, but it's now invalid.
-    // If you want more safety, used the actual c10::intrusive_ptr class
+    if (self && --self->refcount_ == 0) {
```

Comment:
The reason for `incref`/`decref` not actually doing incref/decref is that any logic touching the refcounter is complex, easy to get wrong (especially how refcount and weakcount interact with each other) and should be constrained to as small a surface as possible.

This is why only the `intrusive_ptr` class itself was allowed to touch `refcount_` and `weakcount_` and anything outside (i.e. even the `incref`/`decref` functions) were not allowed to touch it.

I agree though that this made `incref()` a bit more complicated than it would have been otherwise. So I'm ok with changing `incref()` if you think this is better.

For `decref()`, however this change makes things strictly worse. `decref` is now more complicated than it was before, and it is basically a duplication of `intrusive_ptr::reset()`. Can we avoid the duplication somehow? If not, let's just keep it as it was. See also my general comment on why these changes don't change perf and it's probably side-stepping the assertion that causes the perf improvements. So if we use a different way to side-step the assertion, then these changes here aren't needed anymore.

Reply:
You are providing a incref that does incref but in a different way, it's no better than just increment refcount_ inside incref. And you are also relying on compiler to optimize unnecessary code away, which is not reliable and in our case indeed it doesn't do what you expected. You are relying on intrusive_ptr destructor to decref right? Why is this worse?